smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1810 @@
+    LaneOffset as LaneOffsetElement,
+)
+from opendrive2lanelet.opendriveparser.elements.roadLanes import (
+    LaneSection as LaneSectionElement,
+)
+from opendrive2lanelet.opendriveparser.elements.roadLanes import (
+    LaneWidth as LaneWidthElement,
+)
+from opendrive2lanelet.opendriveparser.elements.roadPlanView import (
+    PlanView as PlanViewElement,
+)
+from opendrive2lanelet.opendriveparser.parser import parse_opendrive
+from lxml import etree
+from shapely.geometry import Polygon
+import rtree
+
+from smarts.core.road_map import RoadMap, Waypoint, WaypointsCache
+from smarts.sstudio.types import MapSpec
+from smarts.core.utils.math import (
+    CubicPolynomial,
+    constrain_angle,
+    distance_point_to_polygon,
+    get_linear_segments_for_range,
+    offset_along_shape,
+    position_at_shape_offset,
+    vec_2d,
+    inplace_unwrap,
+    radians_to_vec,
+)
+
+from .lanepoints import LinkedLanePoint, LanePoints
+from .coordinates import BoundingBox, Point, Pose, RefLinePoint, Heading
+from smarts.core.utils.geometry import generate_mesh_from_polygons
+
+
+def _convert_camera(camera):
+    result = {
+        "name": camera.name,
+        "type": "perspective",
+        "perspective": {
+            "aspectRatio": camera.fov[0] / camera.fov[1],
+            "yfov": np.radians(camera.fov[1]),
+            "znear": float(camera.z_near),
+            # HACK: The trimesh gltf export doesn't include a zfar which Panda3D GLB
+            #       loader expects. Here we override to make loading possible.
+            "zfar": float(camera.z_near + 100),
+        },
+    }
+    return result
+
+
+gltf._convert_camera = _convert_camera
+
+
+class _GLBData:
+    def __init__(self, bytes_):
+        self._bytes = bytes_
+
+    def write_glb(self, output_path):
+        with open(output_path, "wb") as f:
+            f.write(self._bytes)
+
+
+@dataclass
+class LaneBoundary:
+    refline: PlanViewElement
+    inner: "LaneBoundary"
+    lane_widths: List[LaneWidthElement]
+    lane_offsets: List[LaneOffsetElement]
+    segment_size: float = 0.5
+
+    def refline_to_linear_segments(self, s_start: float, s_end: float) -> List[float]:
+        s_vals = []
+        geom_start = 0
+        for geom in self.refline._geometries:
+            geom_end = geom_start + geom.length
+            if type(geom) == LineGeometry:
+                s_vals.extend([geom_start, geom_end])
+            else:
+                s_vals.extend(
+                    get_linear_segments_for_range(
+                        geom_start, geom_end, self.segment_size
+                    )
+                )
+            geom_start = geom_start + geom.length
+        return [s for s in s_vals if s_start <= s <= s_end]
+
+    def get_lane_offset(self, s: float) -> float:
+        if len(self.lane_offsets) == 0:
+            return 0
+        i = 0
+        if s < self.lane_offsets[0].start_pos:
+            return 0
```

Comment:
We don't need to compute the lane offset for any value of s before the first `lane_offset` begins since the `inner_boundary` of lane matches the central reference line. But it is guarantee that the last lane_offset will go to till the end of the lane if s is beyond its `start_pos` and hence we need to compute the lane offset for that value, even though that lane_offset may be computed to be 0 (if the lane boundary and central reference line matches in the end).
This is how they have specified it in the [documentation.](https://releases.asam.net/OpenDRIVE/1.6.0/ASAM_OpenDRIVE_BS_V1-6-0.html#_lane_offset)

Reply:
ah, ok, got it.  (I think?)